ocaml/typing/typedecl.ml

```diff
@@ -528,6 +531,79 @@ let verify_unboxed_attr unboxed_attr sdecl =
       end
   end
 
+(* Note [Default layouts in transl_declaration]
+   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   For every type declaration we create in transl_declaration, we must
+   choose the layout to use in the [type_layout] field.
+
+   1. If there is a layout annotation, use that. We might later compute
+      a more precise layout for the type (e.g. [type t : value = int] or
+      [type t : value = A | B | C]); this will be updated in
+      [update_decl_layout], which also ensures that the updated layout is
+      a sublayout of the annotated layout.
+
+   2. If there is no annotation but there is a manifest, use the layout
+      of the manifest.
+
+   3. If there is no annotation and no manifest, the default layout
+      depends on the kind:
+
+      - Abstract types: In this case, we have a fully abstract type declaration,
+        like [type t]. We wish to default these to have layout [value] for
+        backward compatibility.
+
+      - [@@unboxed] records and variants: We use [any] as the default.
+        This default gets updated in [update_decl_layout], when we can
+        safely look up the layout of the field. Recursive uses
+        of the unboxed type are OK, because [update_decl_layout] uses
+        [Ctype.type_layout], which looks through unboxed types (and thus
+        the choice of [any] is not observed on recursive occurrences).
+
+      - Other records and variants: The layout of these depends on the layouts
+        of their fields: an enumeration variant is an [immediate], and someday
+        (* CR layouts v5: today is the someday! *) we will allow records
+        comprising only [void]s, which will also be [immediate].
+
+        So we choose a default of [value], which gets updated in
+        [update_decl_layout]. This default choice does get used when updating
+        the layouts of other types that (recursively) mention the current type,
+        but that's OK: the update in [update_decl_layout] can only change a
+        [value] to become [immediate], and yet that change can never affect
+        the decision of whether an outer record/variant is a [value] or
+        [immediate] (only choices of [void] can do that).
+
+        (Again, any unboxed records/variants are looked through by
+        [type_layout], so a void one of those is OK.)
+
+        It is tempting to use [any] as the default here, but that causes
+        trouble around recursive occurrences in [update_decl_layout].
+
+      - Extensible variants: These really are [value]s, so we just use
+        that as the default.
+
+   There is a notable flaw in this plan, as we see in this example:
```

Comment:
I'm not sure I'd call this a flaw.  Examples like this are why we only promise that the stored layout will be an upper bound, and as you note it's not a problem because the appropriate expansions happen later.  But it's a nice illustration of why we can't promise something stronger - maybe revise the comment slightly to indicate that?